### PR TITLE
Memoize isOnRemoteNode in index settings and refactor as well

### PR DESCRIPTION
**Pull request by gbbafna**
### Description

Memoize isOnRemoteNode in index settings and refactor as well

### Related Issues
Resolves #[Issue number to be closed when this PR is merged]

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] ~Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/IndexSettings.java [hunk 1]
```diff
@@ -984,6 +985,7 @@ public IndexSettings(final IndexMetadata indexMetadata, final Settings nodeSetti
          * Now this sortField (IndexSort) is stored in SegmentInfo and we need to maintain backward compatibility for them.
          */
         widenIndexSortType = IndexMetadata.SETTING_INDEX_VERSION_CREATED.get(settings).before(V_2_7_0);
+        isOnRemoteNode = RemoteStoreNodeAttribute.isRemoteDataAttributePresent(this.getNodeSettings());
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/IndexSettings.java:988]: `assignableOnRemoteNode` seems to appropriate

Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/IndexSettings.java:988]: renaming it to `assignedOnRemoteNode` . Let me know if you think otherwise .

**Review by scrawfor99 — COMMENTED** [review]
Hi @gbbafna, 

Thanks for putting this together.

The changes all look pretty straightforward but I am not sure how much impact this caching will have. I see how you are checking whether the settings is present on the remote node but I don't see where that logic is used.

Since you swapped

```
 public boolean isRemoteNode() {
        return RemoteStoreNodeAttribute.isRemoteDataAttributePresent(this.getNodeSettings());
```
for
```
public boolean isAssignedOnRemoteNode() {
        return assignedOnRemoteNode;
```
It seems like the change of 

```
return ReplicationType.SEGMENT.equals(replicationType) || isRemoteNode();
```
for
```
return ReplicationType.SEGMENT.equals(replicationType) || isAssignedOnRemoteNode();
```
maintains the same behavior but is reading from the variable. Presumably this would show some speed up but it is not clear to me how much of a difference this would make. Is this the extent of the intended speed up or are there further changes you had in mind? 
 
Do you have any testing to compare the results with/without the change?

**Review by ashking94 — APPROVED** [review]
LGTM, this should reduce the unnecessary look ups as well.

**Comment by gbbafna** [comment]
> Is this the extent of the intended speed up or are there further changes you had in mind?

I don't have extent of the intended speed up . This is because remote store migration is a new feature under development and is experimental . We have reduced the unnecessary map look ups , similar to other index settings as well .

File: server/src/main/java/org/opensearch/index/IndexSettings.java [hunk 2]
```diff
@@ -984,6 +985,7 @@ public IndexSettings(final IndexMetadata indexMetadata, final Settings nodeSetti
          * Now this sortField (IndexSort) is stored in SegmentInfo and we need to maintain backward compatibility for them.
          */
         widenIndexSortType = IndexMetadata.SETTING_INDEX_VERSION_CREATED.get(settings).before(V_2_7_0);
+        assignedOnRemoteNode = RemoteStoreNodeAttribute.isRemoteDataAttributePresent(this.getNodeSettings());
```
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/index/IndexSettings.java:988]: `NodeSettings` can't be changed without process restart, right?

**Review by sachinpkale — APPROVED** [review]

File: server/src/main/java/org/opensearch/index/IndexSettings.java (hunk 2 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/IndexSettings.java:988]: Yes, that is only populated from yml only .

**Pull request merged**